Issue408 refrigerants r134a r410a

MirkoEngelpracht opened the pull request:
Hello everybody,
the issue is solved and ready to be reviewed.

I've added a few examples and validation models. Within the validation models, I've checked all medium functions including the fitted formulas as well as the derivatives. Furthermore, I've compared the results to external media libraries, i.e. the HelmholtzMedia library and ExternalMedia library.

Actually, some problems occur while using partial derivatives with respect to specific entropy. Moreover, I'm not quite satisfied with my solution of implementing a general form of two-dimensional polynomials as done, for example, for the function 'density_pT'. I've also added the coefficients of fitted formulas directly in the model for reasons of simulation speed.

Perhaps somebody has some ideas to provide a nicer solution for the issues mentioned above.

Best regards
Mirko

MirkoEngelpracht commented:
Unfortunately, one of the checks failes, i. e. "continous-integration/travis-ci/pr". Can somebody tell what that means and what I've to do to pass the check?

Best regards
Mirko

mlauster commented:
You can simply lick on details nextto the test and you will get to the travis web page. There you can have a look at the log ad see, which problems occur. There are a lot of html warnings, too much to make a screenshot. But this is the direct link to the log

https://travis-ci.org/RWTH-EBC/AixLib/builds/283737785?utm_source=github_status&utm_medium=notification

mlauster commented:
There also seemed some Travis internal issues with loading packages, what happens from time to time. I restarted the job just to make sure we get the full picture

mlauster commented:
I think it's only the number of warnings and the tests take more than 70 seconds, what we set at maximum. Maybe we need to check this threshold again, @MichaMans?

mlauster commented:
@MirkoEngelpracht, are the mentioned problems the ones you mentioned before for closed loop systems or could you solve these ones?

MirkoEngelpracht commented:
@mlauster: First of all, thanks a lot for your quick answer and explanation of Travis. I assume that I've to correct the html errors/warnings. Am I right?

The mentioned problems have nothing to do with the problems concerning close loop systems. For the latter, I've solved the problems by changing my closed loop system (i.e. I added a dummy heat exchanger to pretend some "action" in the refrigerant cycle) and this minor change solved the problem. Additionally, I've tested further close loop systems to be sure that the refrigerant will work for close loop systems. I tested the following models:

- Modelica.Fluid.Examples.BranchingDynamicPipes
- Modelica.Fluid.Examples.HeatingSystem
- AixLib.Fluid.Examples.ResistanceVolumeFlowReversal

I've to addmit that it doesn't make sense to use refrigerants as working fluids in the models mentioned above because the refrigerant often will be and not be liquid. However, it demonstrates that the medium model works for close loop systems.

Nevertheless, problems occur while translation and index reduction since it's not possible to differentiate advanced expressions as "MediumModel.setState_phX" or "MediumModel.setState_pTX". These expressions are used, for example, within the model "AixLib.Fluid.Examples.PumpRadiatorThermostaticValve". The reason for the problem is that the thermodynamic state record (i.e. the output of the advanced expressions) contains an integer (i.e. "phase" indicating two-phase-region or single-phase-region) and it's not possible to differentiate discrete variables. The problem occurs for all two-phase media models and I've checked it for the following media or media libraries:

- Modelica.Media.R134a.R134a_ph
- HelmholtzMedia library
- ExternalMedia library

In conclusion, the problem seems to be a general problem for two-phase media models. In order to provide a proper solution, I've tried to define manually derivatives of these expressions by both excluding "phase" as derivative (i.e. by using the "noDerivative=phase" annotation) and by defining derivatives of all functions called by the setState functions (e.g. density(p,h) or density(state)). As far as I can see, this solution seems to work and I'm able to translate and simulate, for example, the model "AixLib.Fluid.Examples.PumpRadiatorThermostaticValve". I want to point out again that it doesn't make sense to use a refrigerant here as it will always be superheated.

The problems concerning the partial derivatives (i.e. especially with respect to specific entropy) are new problems. To be more specific, the following partial derivatives are not calculated correctly:

- (ds/dT)@d=const, missing calculation rule for two-phase-region
- (du/dd)@T=const, missing calculation rule for two-phase region

- (dd/dp)@s=const, this derivative uses the derivative (ds/dT)@d=const

The former is used to calculate derivatives wrt. time for the following functions:

- T(p,s):	I observed little influence of the partial derivatives mentioned above
- d(p,s):	I observed great influence of the partial derivatives mentioned above, however, mostly in the two-phase region
- h(p,s):	I observed little influence of the partial derivatives mentioned above

In consequence, it's not possible to use the derivatives wrt. time of these three functions by now.
I guess that the problem could be solved by adding calculation rules for the two-phase-region for the following basic derivatives (i.e. derivatives that only depends on derivatives wrt. density or temperature and, therefore, on derivatives of the Helmholtz equation of state):

- (ds/dT)@d=const, missing calculation rule for two-phase-region
- (du/dd)@T=const, missing calculation rule for two-phase region

These calculation rules are the only missing ones and all other calculation rules are validated by comparing them to the HelmholtzMedia library. Perhaps @cvering can help out here?

mlauster commented:
Maybe @thorade has also some thoughts to bring in? I now you were working with media models, so your knowledge is appreciated!

mlauster commented:
And yes, @MirkoEngelpracht, you need to fix them. :-) @MichaMans, did you had some automation scripts for doing this?

thorade commented:
A few things you probably already know:
Closed loops can be difficult, especially with media with temperature dependent density. One possible fix is to introduce a component like a storage tank, e.g. directly behind the condenser. But it seems you have added another component that has a similar effect.

Regarding the problem that setState functions cannot easily be differentiated:
That is the reason why BaseProperties exist, at least to my understanding.
https://stackoverflow.com/questions/11207029/modelica-media-baseproperties-versus-setstate-xxx
Maybe this old discussion contains something helpful?
https://github.com/ibell/coolprop/issues/61

The two derivatives you mentioned in the two-phase region should be simple, I believe:
for du/dT|d see https://github.com/thorade/HelmholtzMedia/blob/master/HelmholtzMedia/Interfaces/PartialHelmholtzMedium/specificEnthalpy_derd_T.mo
and replace h with u. 
For ds/dT|d you might make use of the fact that cv=T * (ds/dT)_v = T*(ds/dT)_d = (du/dT)_v

thorade commented:
Oh, and R410a is a binary mixture, so evaporation is no longer isothermal and the equations in the two phase region are probably different from R134a

MirkoEngelpracht commented:
@mlauster: Well, I saw that coming. I've corrected hopefully all syntax errors and pushed the new branch.
@thorade: Thank you so much for all your hints and help, I'd really appreciate it :)

Regarding R410a: You're absolutely right, R410a is a mixture. However, we assume R410a to behave like a pseudo-pure fluid since its temperature glide is very low with about 0.1 K. We know that R410a still behaves truly zeotropic and should not be handled azeotropic, however, assuming azeotripic behavior should work with respect to my master thesis' aim. Consequently, calculation procedures are only roughly valid within the two-phase region and, therefore, the two-phase region is a limitation of the model.

Regarding close loop systems: Indeed, I've added both a dummy heat exchanger as mentioned above and a dummy storage tank behind the condenser. So the latter helped me to make the model work.

Regarding the missing partial derivatives: Perfect, that really helped me. By adding the two derivatives, all relevant derivatives should be defined now. Thanks a lot!

Regarding the problem of the setState functions: The two discussion are really interesting and I've learned some new things I didn't know about. I'll try to apply some of your thoughts.

mlauster commented:
@thorade, thanks for your input, I knew you would have some interesting links to provide!

mlauster commented on AixLib/Media/Refrigerants/Interfaces/PartialHybridTwoPhaseMediumFormula.mo:
```diff
@@ -0,0 +1,2345 @@
+  end density_derT_p;
+
+  replaceable function temperature_derp_h
+    "Calculates temperature derivative (dT/dp)@h=const."
```
Please only ASCII signs, otherwise Python has problems loading result files

MirkoEngelpracht replied:
I'm going to remove wrong signs.

MirkoEngelpracht commented on AixLib/DataBase/Media/Refrigerants/R290/TSP_IIR_P05_30_T263_343.mo:
```diff
@@ -0,0 +1,70 @@
+within AixLib.DataBase.Media.Refrigerants.R290;
+record TSP_IIR_P05_30_T263_343
+  "Record with fitting coefficients taken from the Fast_Propane model"
+  import AixLib;
```
Well, something went wrong. I didn't intend to do that and I'm going to remove it.

MirkoEngelpracht commented on AixLib/DataBase/Media/Refrigerants/R410a/BDSP_IIR_P1_48_T233_340.mo:
```diff
@@ -0,0 +1,64 @@
+within AixLib.DataBase.Media.Refrigerants.R410a;
+record BDSP_IIR_P1_48_T233_340
+  "Record with fitting coefficients calculated for first implementation"
+  import AixLib;
```
See above.

MirkoEngelpracht commented on AixLib/Media/Refrigerants/Examples/RefrigerantTestImplementationB.mo:
```diff
@@ -0,0 +1,33 @@
+within AixLib.Media.Refrigerants.Examples;
```
The difference is very little. Both models are standard models provided by the Modelica standard library. I'm going to remove Example A.

MirkoEngelpracht commented on AixLib/Media/Refrigerants/Examples/RefrigerantTestImplementationB.mo:
```diff
@@ -0,0 +1,33 @@
+model RefrigerantTestImplementationB
+  "Model to test the refrigerant implementation"
+  extends Modelica.Icons.Example;
+  extends Modelica.Media.Examples.Tests.Components.PartialTestModel2(
```
See above.

MirkoEngelpracht commented:
Thanks a lot for your first remarks, @mlauster!

1. I've removed the import commands.
2. I've removed the example model A since the difference between model A and B is little. Both models are introduced in the Modelica standard library (Modelica.Media.Examples.Tests.Components.*) and are basic models to test media models. The second one, Example B, is slightly larger then Example A and consists of the following elements: Fixed Mass Flow - Volume - Pipe - Volume - Pipe - Boundary. 
3. Hopefully, I've romoved all non-ASCII signs. Please let me know if I forgot some signs. If there is any possibility to search automatically for signs that are not ASCII with Atom?

4. As requested, I've shortened all lines except for the HTML links.
5. I totally agree with you concerning the Users Guide. Therefore, I'll work on it at the weekend.

6. As discussed, @cve, I've changed the calculation rules of transport properties for the two-phase region. Now, the lever rule is applied.

mlauster commented:
Thanks a lot, @MirkoEngelpracht, please let me know when you are done with your changes.

MirkoEngelpracht commented:
@mlauster: I'm done with the changes requested so far.

mlauster commented on AixLib/Media/Refrigerants/Examples/RefrigerantProperties.mo:
```diff
@@ -0,0 +1,358 @@
+    "Record that contains further thermodynamic properties calculated during
+    the test";
+  StateErrors stateErros
+    "Record that contains the the errors occuring using the different
```
one "the" too much

mlauster commented:
Hmm, I found 44 errors and related 43 warnings all relating to the validation package. It seems to be mostly about functions without explicit declaration. Maybe this is due to removing "import AixLib"-statements without extending the pathes? However, I assume they might be easy to fix. Or are these well known and long-standing issues? It would be really nice to have no warnings and errors at least in "normal" code checking mode.

![unbenanntes bild](https://user-images.githubusercontent.com/6892690/31684123-11bb0f80-b37f-11e7-8c14-30e1dda167e8.png)

mlauster commented:
Checking in pedantic mode reveals some further errors, mainly related to annotations. They can often be easily fixed by using an alternative construct or by removing them. :)

![unbenanntes bild1](https://user-images.githubusercontent.com/6892690/31684497-4addd238-b380-11e7-8811-fc44c91f95f2.png)
![unbenanntes bild2](https://user-images.githubusercontent.com/6892690/31684498-4b175e0e-b380-11e7-90e2-080bb5922d8b.png)
![unbenanntes bild3](https://user-images.githubusercontent.com/6892690/31684499-4b4986ae-b380-11e7-8ef1-852f7d60bb78.png)

MirkoEngelpracht commented:
@mlauster: Yeah, I see that and I'd really appreciate it :) Thanks al lot for all your help and feedback!

**Regarding the 44 errors and related 43 warning**
Ouh I see, these errors occur because I use the HelmholtzMedia-library for validation (i.e. I use the library within the validation models "RefrigerantsFittedFormulas" and "RefrigerantDerivatives"). I guess that you don't have included the library?

Well, the usage of the HelmholtzMedia-library seems then to be a general problem. How shall we proceed here? Shall I delete these two validation models? To be honest, I don't like this idea since the HelmholtzMedia-library is a very nice library and perfectly suitable for validation. Or shall I just add a comment within the information section that the library is required compulsory to run the validation models?

**Regarding the pedantic errors**
Hopefully, I've corrected all errors occurring due to pedantic check.

mlauster commented:
@MirkoEngelpracht, could you run the Helmholtz media validation cases and save the parts of results that come from Helmholtz media into a table (or even better, include them as table in the model within a CombiTimeTable, with "onFile=false", see e.g. VDI test cases of thermal zones)? In this way, one could still validate the models without having these dependencies. And of course, you need to mention where the results come from in your documentation. :-)

MirkoEngelpracht commented:
@mlauster, thanks for your hint :) As proposed, I've added CombiTimeTables for the validation models. Therefore, these models should now be executable.

mlauster commented:
@MirkoEngelpracht, we're getting close! :-) As one last thing, in pedantic mode I get six errors, all complaining about the use of the unknown (because vendor specific) annotation `__Dymola_experimentFlags`. Do you need that flag or could you get rid of it?
All occasions are in validation and examples, `RefrigerantProperties`,  `RefrigerantTestImplementation` (by the way, that sounds like an old model you used for testing, is there a better name and is it useful?), `RefrigerantsFittedFormulasR134a`, `RefrigerantsFittedFormulasR290`, `RefrigerantsFittedFormulasR410a`, `RefrigerantsInversions`.
The advantage of passing pedantic check is that the code will as well run in open-source environments such as OpenModelica and JModelica.

MirkoEngelpracht commented:
@mlauster, I corrected the errors you mentioned. 

However, how do you conduct the pedantic check? If I conduct the pedantic check by activating the flag (Simulation Tab -> Simulation Setup -> Translation -> Checkbox 'Pedantic mode for checking Modelica semantics'), I won't see the errors you mentioned. I use Dymola 2018. Perhaps you know the reason or can give me some hints?

mlauster commented:
Hm, that's similar to what I do. But I still use Dymola 2017 (shame on me).

By the way, everything is fine now! :) Thanks a lot for these models!

MirkoEngelpracht commented:
@mlauster, well, that's odd regarding the pedantic check but, luckily, you could told me the errors! And thank you very much for all your reviews and suggestions for improvement - Even in your spare time! :)

I'll wait with merging this branch into the development branch until @cvering will finish his review as well.

mlauster commented:
@cvering, can we finally merge that? I would like to get the library on a current state
